File: Common/Graph/src/main/java/com/greentree/common/graph/Tree.java
package com.greentree.common.graph;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import com.greentree.common.graph.algorithm.brige.BridgeFinder;
import com.greentree.common.graph.algorithm.component.ConnectivityComponentsFinder;
import com.greentree.common.graph.algorithm.cycle.CycleFinder;
import com.greentree.common.graph.algorithm.cycle.VertexCycle;
import com.greentree.common.util.iterator.IteratorUtil;

public interface Tree<V> extends Graph<V>, CycleFinder<V>, BridgeFinder<V> {
	
	@Override
	default Iterable<? extends DirectedArc<V>> getBridges() {
		return IteratorUtil.empty();
	}
	
	@Override
	default Iterable<? extends VertexCycle<V>> getCycles() {
		return IteratorUtil.empty();
	}
	
	@Override
	default CycleFinder<V> getCycleFinder() {
		return this;
	}
	
	@Override
	default BridgeFinder<V> getBridgeFinder() {
		return this;
	}
	
	@Override
	default ConnectivityComponentsFinder<V> getConnectivityComponentsFinder() {
		return new TreeConnectivityComponentsFinder<>(this);
	}
	
	public static final class TreeConnectivityComponentsFinder<V>
			implements ConnectivityComponentsFinder<V> {
		
		private final Tree<? extends V> tree;
		
		public TreeConnectivityComponentsFinder(Tree<? extends V> tree) {
			this.tree = tree;
		}
		
		@Override
		public int getComponent(V vertex) {
			if(tree.has(vertex))
				return 0;
			throw new IllegalArgumentException("not vertex");
		}
		
		@Override
		public List<? extends Collection<V>> getComponents() {
			final var result = new ArrayList<Collection<V>>();
			final var c = new ArrayList<V>();
			result.add(c);
			for(var v : tree)
				c.add(v);
			return result;
		}
		
	}
	
}

File: Commons/Graph/src/main/java/com/greentree/commons/graph/Tree.kt
package com.greentree.commons.graph

import com.greentree.commons.graph.algorithm.brige.BridgeFinder
import com.greentree.commons.graph.algorithm.component.ConnectivityComponentsFinder
import com.greentree.commons.graph.algorithm.cycle.CycleFinder
import com.greentree.commons.graph.algorithm.cycle.VertexCycle
import com.greentree.commons.util.collection.EmptyCollection
import com.greentree.commons.util.iterator.IteratorUtil

interface Tree<out V : Any> : FiniteGraph<V>, CycleFinder<V>, BridgeFinder<V> {

	override val bridges: Iterable<DirectedArc<V>>
		get() = IteratorUtil.empty()
	override val cycles: Collection<VertexCycle<V>>
		get() = EmptyCollection()
	override val bridgeFinder: BridgeFinder<V>
		get() = this
	override val cycleFinder: CycleFinder<V>
		get() = this
	override val connectivityComponentsFinder: ConnectivityComponentsFinder<V>
		get() = TreeConnectivityComponentsFinder(this)

	class TreeConnectivityComponentsFinder<out V : Any>(tree: Tree<V>) : ConnectivityComponentsFinder<V> {

		private val tree: Tree<V> = tree

		override fun getComponent(vertex: @UnsafeVariance V): Int {
			if(vertex in tree) return 0
			throw IllegalArgumentException("not vertex")
		}

		override val components: List<Collection<V>>
			get() {
				val result = ArrayList<Collection<V>>()
				val c = ArrayList<V>()
				result.add(c)
				for(v in tree) c.add(v)
				return result
			}
	}
}
